crates/viewer/re_renderer/src/resource_managers/chroma_subsampling_converter.rs

```diff
@@ -0,0 +1,336 @@
+use smallvec::smallvec;
+
+use crate::{
+    allocator::create_and_fill_uniform_buffer,
+    include_shader_module,
+    renderer::{screen_triangle_vertex_shader, DrawData, DrawError, Renderer},
+    wgpu_resources::{
+        BindGroupDesc, BindGroupEntry, BindGroupLayoutDesc, GpuBindGroup, GpuBindGroupLayoutHandle,
+        GpuRenderPipelineHandle, GpuTexture, PipelineLayoutDesc, RenderPipelineDesc, TextureDesc,
+    },
+    DebugLabel, RenderContext,
+};
+
+use super::ColorPrimaries;
+
+/// Supported chroma subsampling input formats.
+///
+/// Keep indices in sync with `chroma_subsampling_converter.wgsl`
+#[allow(non_camel_case_types)]
+#[derive(Clone, Copy, Debug)]
+pub enum ChromaSubsamplingPixelFormat {
+    /// 4:2:0 subsampling with a separate Y plane, followed by a UV plane.
+    ///
+    /// Expects single channel texture format.
+    ///
+    /// First comes entire image in Y in one plane,
+    /// followed by a plane with interleaved lines ordered as U0, V0, U1, V1, etc.
+    ///
+    ///          width
+    ///          __________
+    ///          |         |
+    /// height   |    Y    |
+    ///          |         |
+    ///          |_________|
+    /// height/2 | U,V,U,… |
+    ///          |_________|
+    Y_UV12 = 0,
+
+    /// YUV 4:2:2 subsampling, single plane.
+    ///
+    /// Expects single channel texture format.
+    ///
+    /// The order of the channels is Y0, U0, Y1, V0, all in the same plane.
+    ///
+    ///             width * 2
+    ///        __________________
+    ///        |                 |
+    /// height | Y0, U0, Y1, V0… |
+    ///        |_________________|
+    ///
+    YUYV16 = 1,
+}
+
+impl ChromaSubsamplingPixelFormat {
+    pub fn expected_data_width_height(&self, width: u32, height: u32) -> (u32, u32) {
```

Comment:
This could use a docstring and/or better names to make is 1000% clear that the input is the "decoded size" and the returned size is the "encoded and concatenated size", or something. Names are hard.